text/000-sfc-and-template-import-primitives.md

```diff
@@ -0,0 +1,875 @@
+highly-requested features (such as single-file components) via stable, public
+API.
+
+## Motivation
+
+This proposal is intended to unlock experimentation around two
+highly-requested features:
+
+1. Single-file components.
+2. Module imports in component templates.
+
+Although exploring these features is the primary motivation, an important
+benefit of stabilizing low-level APIs is that they enable the Ember community
+to experiment with new, unexpected ideas.
+
+### Single-File Components
+
+In Ember components today, JavaScript code lives in a `.js` file and template
+code lives in a separate `.hbs` file. Juggling between these two files adds
+friction to the developer experience.
+
+Template and JavaScript code are inherently coupled, and changes to one are
+often accompanied by changes to the other. Separating them provides little
+value in terms of improving reusability or composability.
+
+Other component APIs eliminate this friction in different ways. React uses
+JSX, which produces JavaScript values:
+
+```js
+// MyComponent.jsx
+
+export default class extends Component {
+  state = { name: 'World' };
+
+  render() {
+    return <div>Hello {this.state.name}!</div>
+  }
+}
+```
+
+Vue has an optional [`.vue` single-file component (SFC)
+format](https://vuejs.org/v2/guide/single-file-components.html):
+
+```html
+<!-- MyComponent.vue -->
+
+<template>
+  <div>Hello {{name}}!</div>
+</template>
+
+<script>
+export default {
+  data() {
+    return {
+      name: 'World'
+    }
+  }
+}
+</script>
+```
+
+### Module Imports in Templates
+
+One benefit of JSX is that it leverages JavaScript's existing scope
+system. React components are JavaScript values that can be imported and
+referenced like any other JavaScript value. If a binding would be in scope
+in JavaScript, it's in scope in JSX:
+
+```js
+import { Component } from 'react';
+import OtherComponent from './OtherComponent';
+
+export default class extends Component {
+  render() {
+    return (
+      <div>
+        // We know OtherComponent is in scope because it was imported at the top
+        // of the file, so we can invoke it here as a component.
+        <OtherComponent />
+      </div>
+    );
+  }
+}
+```
+
+This is a bit nicer than Vue's SFCs, where JavaScript code and template code
+happen to be in the same file but otherwise interact no differently than if
+they were still in separate files. Developers must learn a proprietary Vue
+API for explicitly copying values from JavaScript's scope into the template's
+scope, and component names can be different between where they are imported
+(`OtherComponent`) and where they are invoked (`other-component`):
+
+```html
+<template>
+  <div>
+    <other-component />
+  </div>
+</template>
+
+<script>
+import OtherComponent from './OtherComponent.vue'
+
+export default {
+  components: {
+    OtherComponent
+  }
+}
+</script>
+```
+
+Neither of these approaches is exactly right for Ember. Ideally, we'd find a
```

Comment:
Why are neither of these exactly right for Ember? Do you mean that neither of them is exactly aligned with your ideal solution in a general sense? Or do you mean that there is something specifically different about Ember that renders the Vue and React strategies undesirable?